### ISSUE TITLE
opening new My Tracks very slow with Android 12

### ISSUE THREAD

**Issue by pebogufi** [osmandapp/OsmAnd#17805]
## Description
With new My tracks my phone S10 is very slow, but not my Tab with Android 10.
I use  two devices, which are nearly complete identicaly in OsmAnd folders and files.
+Samsung S 10, Android 12, installed on internal, Android/media
+Samsung Tab S5e, Android 10, installed on Sd card, Android/obb
The S10 is normally **much** faster, example: install of osmand 
S10 needs 5-8 sec, 
Tab needs more than 20 sec.

BUT opening my tracks needs (total files in tracks subfolders ~500)
S10 5-7 sec
Tab ~1 sec.
I suppose the problem is the file access restrictions in Android 12 which uses extended file acces when creating statistics (tracks, distance, uphill, downhill, duration, size) which are at bottom line in tracks page.

### proposed change
Please add an option/checkbox to decide whether this automatic calculation of statstics should be done. I think this will not only help a lot of users with new Android versions > 10, but make them really happy !

**Problem 2**
Opening My tracks / "Visible on map"  with S10, Android 12 needs so much time, that i often get a message from system, whether to continue or abort the app.
With Tab it is also very slow, 5-10 sec, but no system warning. Visible files with both devices max 5, total files in tracks subfolders ~500.

**Environment**
OsmAnd Version: OsmAnd~ 4.6.0#537m, veröffentlicht: 2023-08-02
Android version: 12
Device model: Samsung S 10

**Comment by pebogufi** [comment]
With S10:
After removing all but 50 files in tracks subfolders and no tracks visible the function My tracks / "Visible on map" needs 3-5 sec to show only one line "currently recording track" while recording is off.

**Comment by Corwin-Kh** [comment]
This issue should be checked on latest build as it could be already fixed. Aside from this issue, there was another one that is related to SmartFolders and affected opening My places -> tracks. It is fixed in [#2159](https://github.com/osmandapp/OsmAnd-issues/issues/2159)

**Comment by pebogufi** [comment]
> tracks. It is fixed in [#2159](https://github.com/osmandapp/OsmAnd-issues/issues/2159)

Link reports error 404, not found

After using again my folders with 767 gpx files
opening track folder needs 10 sec, that is no improvement to the initial report.

**Comment by Corwin-Kh** [comment]
sorry for link - turn put it's closed. https://github.com/osmandapp/OsmAnd/pull/18393/commits/1994f233989f574aac20df91634974c4f4a2e4fb
this one should work. The case i'm talking about is fixed in this pull request. It is already in last builds. 
So, if it is still present then problem somewhere else. we'll try to find out.

**Comment by sonora** [comment]
Unfortunately, I have the same experience, there is no improvement with latest builds. It is still very slow to open either the My Tracks or the Visible Tracks screen. Something we do there seems to be extraordinarily resource intensive. (Interestingly, while it is always slow, times can vary from 3 to 15 sec, Android 11, ~3500 tracks in the folder).

What's worse is #18077, which is a showstopper for me using our 4.6.x releases at all, I am forced to stay with 4.5 for all my work.

**Comment by vshcherb** [comment]
Refactoring is needed with GPX Database.
1. We need to generate all long SQL queries from Enum of GPX characterestics and GPX setting.
2. We shouldn't read file system list and only get track items GPXDatabase. GPXDatabase should be updated in background first time My Places open or Configure Map > Tracks. On Start we should just read GPXDatabase.
3. Filters code should be simplified as we shouldn't have separate classes for filters - they don't require any specific functions. Instead we should have Conversion function (1 per unit), GPXDatabase should provide generic API to retrieve min/max field from database.
4. GPXDatabase should have automatic regeneration code for characteristic when new characterestic is added or when schema is changed. For that purpose we can store timestamp of generation. Scheme will be defined simply as date (in Android there is API to check existing column).

**Comment by sonora** [comment]
I wonder how we did it with the old, expandable list view. That one was comparatively lightning- fast...

**Comment by sonora** [comment]
Fun fact: When you create a smart folder with filter condition "Visible on map", that one opens essentially instananeously. Opening the built-in "Visible tracks" screen creates the long wait, every time.

**Comment by pebogufi** [comment]
@sonora  VERY fun fact!
Thanks a lot for that hint, it is great.
I can confirm, that it opens instantaneously and 
build in quite slow every time.

**Comment by Chumva** [comment]
Work is still in progress, probably this week there will be UI fixes that should improve speed with my places - tracks screen

**Comment by Chumva** [comment]
@sonora can you try new nightly? There was an added progress update for loading tracks similar to the one that was before the changes with these screens

**Comment by sonora** [comment]
Yes  much better. Roughly halves the times to fully display the My Places > Tracks and Configure map > Tracks screens.

But I also managed to now produce the enclosed crash:
[exception.log.txt](https://github.com/osmandapp/OsmAnd/files/13721121/exception.log.txt).

**Comment by sonora** [comment]
@Chumva , @vshcherb Let me re-open this, the crash is quite frequent, so we should not leave it like that...

**Comment by Chumva** [comment]
@sonora Do you have this crash in latest nightly?

**Comment by sonora** [comment]
@Chumva I just noticed it's a different one: [exception.log.txt](https://github.com/osmandapp/OsmAnd/files/13746081/exception.log.txt) (And it is less frequent, only when visiting the Tracks screen repeatedly.)

Unfortunately, I am also back to about 12+ seconds every time I open the screen ... ;)

**Comment by vshcherb** [comment]
@sonora did you test it on nighhlty or release build? On my phone startup on nightly takes 16 seconds but release build just 3 seconds (installed via gplay)

**Comment by sonora** [comment]
I have buit from our master branch, on 2023-12-21 (evening).

**Comment by sonora** [comment]
Here's another crash: 
[exception.log2.txt](https://github.com/osmandapp/OsmAnd/files/13757064/exception.log2.txt)

**Comment by sonora** [comment]
Also: Possibly connected to the commits here, I am again losing the selected track sorting on occasion. I have my track screen sorted A-Z, and had lately several times found it showing to be sorted by "Last modified", am 100% sure I have not set that myself. May happen when updating the app, or on sone other occasion, not sure.

Can open another issue, if you think ithat's better than noting that here.

**Comment by vshcherb** [comment]
This issue will be closed cause the fixes are complete in 4.6, so next issue could continue in 4.7

**Comment by sonora** [comment]
New crash... [exception.log.txt](https://github.com/osmandapp/OsmAnd/files/13777175/exception.log.txt)

**Comment by vshcherb** [comment]
Unfortunately we had a bad code practice... https://github.com/osmandapp/OsmAnd/commit/5086fb4d081ec17d4ae0561f8cb044d8630fcc36 Still fixing it.
Checked the code different places, please let me know if it happens again

**Comment by pebogufi** [comment]
> New crash... [exception.log.txt](https://github.com/osmandapp/OsmAnd/files/13777175/exception.log.txt)

see also
https://github.com/osmandapp/OsmAnd/issues/18828

**Comment by sonora** [comment]
Still crashing, even with https://github.com/osmandapp/OsmAnd/commit/5086fb4d081ec17d4ae0561f8cb044d8630fcc36: New [exception.log.txt](https://github.com/osmandapp/OsmAnd/files/13778538/exception.log.txt)

**Comment by vshcherb** [comment]
Fixed https://github.com/osmandapp/OsmAnd/commit/7a362ff331415430afa18fc0becae8b128516abf

**Comment by sonora** [comment]
At least no more crashes so far, thanks!

**_But it is still soo slow..._** 11to 14 seconds to open the  My tracks screen. And the same again when from there you tap on "Visible on map".

Somehow what we do there does not scale well at all. It seems to somehow scale with the number of tracks in the device, why? It is hardly acceptable that a screen to display a list of 68 folders takes so long to come up. And again so long, to display a screen of visible tracks, even if _zero_ tracks are displayed on the map??

**Comment by vshcherb** [comment]
Do you test only nightly builds or do you have release builds?

**Comment by vshcherb** [comment]
I mentioned in another issue that I see for last 6 months tremendous difference between builds and I can't improve debug build anyhow with adb package compile...

**Comment by sonora** [comment]
Hm, I am use the master branch to build, turn all debugging off, use no debug libs, and turn all optimizations on, so it should be like a "release". I can explicitly download our 4.6.11 release, release build, if you want.

**Comment by vshcherb** [comment]
I didn't test release builds itself with https://source.android.com/docs/core/runtime/jit-compiler adb shell cmd package compile -m speed -f my-package

**Comment by vshcherb** [comment]
I can see that build is slow by just enabling /disabling pev plugin in plugins section

**Comment by sonora** [comment]
I find no systematic correlation between plugins being on or off. But it is weired that the time it takes to open the screen varies so much, I get everything between 8 and 30 seconds, most of the time between 10 and 14, though. Perhaps a memory issue?

**Comment by vshcherb** [comment]
Well I don't think you've got my point. Obviously every program should scale linearly to the number of files, the issue I think is how slow 1 unique operation perform.

You need to try https://download.osmand.net/releases/net.osmand-4.6.11-4611.apk (and probably compile it extra with ```
adb shell cmd package compile -m speed -f net.osmand```)

**Comment by sonora** [comment]
I have just now tried that build as is, and it is equally slow. Upon the first starts (after placing the tracks in the ttrack folder) Android even complained several times that the app has stalled, and offered to kill it (the app was frozen at the time, ptobably creating the track db).

Afterwards, this build performed no better than the nightly.

Regarding 
>Obviously every program should scale linearly to the number of files,

Absolutrly not! My Windows PC has 300,000 files. If Microsoft followed that scaling law, I could never even start the File Explorer to display the hard drive's root... :wink: 

And my Android phone surely also has thousands of files - but the Samsung My Files app displays every folder level instantaneously.

Reading the entire tracks db (if that's what we do?) every time the user opens the My Places>Tracks screen and then again the "Visible on map" screen seems just the worng strategy.

**Comment by vshcherb** [comment]
With nightly build that starts up for about 10-15 seconds, My places Tracks with 531 Track, distance 42 000 km and size 122.8 MB - still opens for 1 second. So in that case I can't confirm this issue. 
RAM Memory footprint - 90 MB.

**Comment by sonora** [comment]
Yes, the app startup time is about 12 sec for me. I think it has increased some lately (from perhaps half that), but that's not my primary issue.

Can you please try this: Use a file manager app to duplicate all your subfolders in your tracks folder, do that a few times to get to perhaps 4000 files. When the file manager asks to overwrite the existing entries, chose "Rename" or similar so you really create duplicate folders.

I have just tried this:
- For 10 folders with about 500 gpx tracks total, the "Tracks screen display time" is like 2 seconds or so.
- But with the duplication to 80 folders / 4000 tracks, I am up to about 30 seconds loading time...

**Comment by sonora** [comment]
Is there any activity and hope to still fix this regression?

If not I will have to start compiling and publishing some instructions to offload tracks, so that OsmAnd still remains usable for those users of my scientific community who conventionally have several thousand track files in their folder structure and the app keeps stalling all the time.

**Comment by vshcherb** [comment]
There is always a hope to fix though issue not reproduced. No video / no logging, so there is no progress here, only hope that 1 day we will see something. 

I think we should focus to get clear logs + video on synthetic tests, in this conversation there is no specific at all. S10 of course is not a target device any more. But if we reproduce then descision could be made.

**Comment by sonora** [comment]
@vshcherb Ah, perhaps a misunderstanding, Victor, my apologies! When you had re-opened the issue after my post https://github.com/osmandapp/OsmAnd/issues/17805#issuecomment-1872015041 above, I was under the impression you had successfully re-produced it with the steps I had noted there.

Well, I am a little at loss what could help here: a video showing a 12 sec wait before the tracks screen shows content does not seem very helpful? And regarding log files I a not sure what to produce using what mechanism, maybe someone who can reproduce can help out or explain?

**Comment by sonora** [comment]
PS: The workaround seems actually quite clear and we could document it somewhere: **You have to limit the amount of data in the sub-tree of your tracks data folder**. I would still have to research the limits, that's why I am asking - it is likely a certain number of tracks (like <2000 or something), or it could be by number of accumulated tranckpoints...

**Comment by sonora** [comment]
PS2: One side effect we could probably fix fast is to _disable the pull-down refresh_.

If you view a longer list of line items (folders or tracks), and scroll it down, then up again, reaching the top of the list initiates a refresh, it seems, and the screen is again stuck for 12 sec for me every such time.

I suugest initiating a refresh when pulling down while you have reached the top of a list on this screen is a feature not really needed here, so we could disable that?

**Comment by sonora** [comment]
I belive that this is the culprit and would need checking: https://github.com/osmandapp/OsmAnd/blob/a763ecc666c0a65ca21483b83956c576995e5cc0/OsmAnd/src/net/osmand/plus/myplaces/tracks/TrackFoldersHelper.java#L124

**Comment by sonora** [comment]
PS: A considerable portion of the wait time (about half) is spent for this line, **even if I have no Smart Folders created**;
https://github.com/osmandapp/OsmAnd/blob/5ca0048428c1d16821eae8b089fde852265fdecb/OsmAnd/src/net/osmand/plus/configmap/tracks/TrackFolderLoaderTask.java#L95

**Comment by sonora** [comment]
This obvious fix at least brings some noticeable improvement whenever no smart folders are present: #19228.

**Comment by sonora** [comment]
And also: Isn't this sorting code https://github.com/osmandapp/OsmAnd/blob/5ca0048428c1d16821eae8b089fde852265fdecb/OsmAnd/src/net/osmand/plus/track/helpers/GPXFolderUtils.java#L30C1-L45C2 superseded by _TracksAdapter.java_?

**Comment by vshcherb** [comment]
> (1) The time to open the Tracks screen should not depend on the total number of tracks in all folders, but merely on the number of items to be shown on the screen to be opened. (In my case simply 70 subfolders and zero tracks on the top level screen)

It's not correct ... Time depends on the number of tracks that's totally fine. Loading & sorting each time your open a screen is fine as well to create a structure especially on new load.

What's not fine that 3500 is a very small number! Loading 3500 records from db and sorting 3500 numbers is about 0.1 seconds should be. 
```
270 t/s - 3500 tracks: 12 - 14 sec
300 t/s - 2500 tracks: 7 - 8 sec
~500 t/s - 1500 tracks: 3 sec
```
It looks totally fine that it's linear dependency but 0.2 millisec per track is too much. Did you try release versions ?  https://download.osmand.net/releases/net.osmand-4.7.2.apk

**Comment by sonora** [comment]
@vshcherb 
- Can we please merge this PR https://github.com/osmandapp/OsmAnd/pull/19228? It makes sense in any case, speeds things up by ~50% (if you don#t have a smart folder), saves CPU cycles and battery power.
- Sorting **every** time: Well, the sorting is important for what you see on the screen, not what's in the background list... :wink: I almost never look at the "All" list, the only one which would require sorting of all 3500 line items.
- Time stamps: see e.g this log file (this already includes the improvements by https://github.com/osmandapp/OsmAnd/pull/19228?): 
[logcat.log.txt](https://github.com/osmandapp/OsmAnd/files/14481053/logcat.log.txt). It implies like 4 seconds, e.g. 

```
03-03 18:13:52.584   810  2715 I net.osmand: TrackFolderLoaderTask Start loading tracks in tracks {pool-20-thread-1}
03-03 18:13:56.668   810  2715 I net.osmand: TrackFolderLoaderTask Finished loading tracks. took 4085ms {pool--20-thread-1}
```
But the actual time I am waiting for the list to appear on screens seems even longer (often like twice that), not sure where the extra time is spent. (PS: I find that using a pull-down refresh on the Tracks screen often is quicker than when I re-enter the screen from the menu, also that seems weird.) an, of course, as stated above, I believe there is no need to again re-load the very same the tracks list when changing to the "Visible on maps" screen.

And yes, you are right, reading a file list table of 3500 files should be peanuts, that's what I am saying all along ... ;)

**Comment by vshcherb** [comment]
1. That's up to @Chumva  I think 

Tonight (or later this week) we decided finally switch nightly builds to release builds. We just tested that startup time is 5 times faster on release builds

**Comment by sonora** [comment]
@Chumva Some more code questions:

- https://github.com/osmandapp/OsmAnd/blob/39185d463013f6e328f851d62bbb4b655a8525bc/OsmAnd/src/net/osmand/plus/myplaces/tracks/filters/SmartFolderHelper.kt#L248

  - Should simply return `allAvailableTrackItems`, it already is class `HashSet`?

- https://github.com/osmandapp/OsmAnd/blob/39185d463013f6e328f851d62bbb4b655a8525bc/OsmAnd/src/net/osmand/plus/myplaces/tracks/filters/SmartFolderHelper.kt#L201C1-L204C1
  - Same here. And why detour via `newSet` and not directly via `allAvailableTrackItems.add(item)`?

**Comment by sonora** [comment]
With this I have the best experience yet, times are around 5 sec without a smart folder and 12 sec with: https://github.com/osmandapp/OsmAnd/pull/19239 (may want to remove the commented lines if merging).

**Comment by sonora** [comment]
In addition to the above optimization PR, here is another quite interesting observation:

I used to have my ~3500 track files stored in 70 subfolders directly under OsmAnd's tracks root folder. Now I have moved my own folders (i.e. everything but rec, import, travel) under one ininterim "my-tracks" folder. That again halves the time to show the MyPlaces/Tracks screen (and tapping it to get to my own subfolders is anyway instantaneous then ...)

**Comment by vshcherb** [comment]
Today was new build released that's not debug and signed as release apk... So everything needs to be reinstalled actually. I tested it still doesn't work as fast as release but this one became 20-50% faster. 
So I think to let you test the new builds and close the issue. Actually the only extra test we could do again is to load 3000 tracks as @sonora mentionned

**Comment by vshcherb** [comment]
Actually I was wrong ... Now is build officially released, so it's ready to test.

**Comment by vshcherb** [comment]
yes I can confirm current nightly build 3x-5x faster and a lot faster actually

**Comment by vshcherb** [comment]
Yes assigned to review & merge

**Comment by sonora** [comment]
@vshcherb Just out of curiosity: Using e.g. _assembleAndroidFullOpenglFat**Release**_ should produce a "fast" build, no matter if we sign it with our debug or our release key, or am I missing anything here?

**Comment by sonora** [comment]
Small success story here: After quite some work on streamlining the code, removing recursions, and removing surplus track and folder analysis calculations, @Chumva and I were able to reduce the time it takes to load the 'MyPlaces>Tracks' screen by an order of magnitude. On my setup from one ~40 sec to now <3 sec.

And also the logcats now reflect significant improvement:
```
05-16 01:34:39.260 31493 32007 I net.osmand: GpxDbHelper Time to loadGpxItems 4173 ms, 3445 items {Initializing app}
05-16 01:34:39.288 31493 32007 I net.osmand: GpxDbHelper Time to loadGpxDirItems 28 ms items count 3445 {Initializing app}
05-16 01:34:39.288 31493 32007 I System.out: Initialized GPX_DB_INITIALIZED in 4201 ms
05-16 01:34:39.367 31493 32007 I System.out: Initialized POI_FILTERS_INITIALIZED in 79 ms
```
